apis/alloydb/v1beta1/instance_identity.go

```diff
@@ -0,0 +1,123 @@
+
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/apis/common"
+	refsv1beta1 "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const serviceDomain = "//alloydb.googleapis.com"
+
+// InstanceIdentity defines the resource reference to AlloyDBInstance, which "External" field
+// holds the GCP identifier for the KRM object.
+type InstanceIdentity struct {
+	parent *InstanceParent
+	id     string
+}
+
+func (i *InstanceIdentity) String() string {
+	return i.parent.String() + "/instances/" + i.id
+}
+
+func (i *InstanceIdentity) ID() string {
+	return i.id
+}
+
+func (i *InstanceIdentity) Parent() *InstanceParent {
+	return i.parent
+}
+
+// AsExternalRef builds a externalRef from a PrivilegedAccessManagerEntitlement.
+func (i *InstanceIdentity) AsExternalRef() *string {
+	er := serviceDomain + "/" + i.String()
+	return &er
+}
+
+type InstanceParent struct {
+	clusterName string
+}
+
+func (p *InstanceParent) String() string {
+	return p.clusterName
+}
+
+// New builds a InstanceIdentity from the Config Connector Instance object.
+func NewInstanceIdentity(ctx context.Context, reader client.Reader, obj *AlloyDBInstance) (*InstanceIdentity, error) {
+
+	// Get Parent
+	clusterRef, err := refsv1beta1.ResolveAlloyDBCluster(ctx, reader, obj, obj.Spec.ClusterRef)
+	if err != nil {
+		return nil, fmt.Errorf("cannot resolve AlloyDBCluster ref: %w", err)
+	}
+
+	// Get desired ID
+	resourceID := common.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		resourceID = obj.GetName()
+	}
+	if resourceID == "" {
+		return nil, fmt.Errorf("cannot resolve resource ID")
+	}
+
+	// Use approved ExternalRef
+	externalRef := common.ValueOf(obj.Status.ExternalRef)
+	if externalRef != "" {
+		// Validate desired with actual
+		actualParent, actualResourceID, err := ParseInstanceExternalRef(externalRef)
+		if err != nil {
+			return nil, err
+		}
+		if actualParent.clusterName != clusterRef.String() {
+			return nil, fmt.Errorf("spec.clusterRef changed, expect %s, got %s", actualParent.clusterName, clusterRef)
+		}
+		if actualResourceID != resourceID {
+			return nil, fmt.Errorf("cannot reset `metadata.name` or `spec.resourceID` to %s, since it has already assigned to %s",
+				resourceID, actualResourceID)
+		}
+	}
+	return &InstanceIdentity{
+		parent: &InstanceParent{
+			clusterName: clusterRef.String(),
+		},
+		id: resourceID,
+	}, nil
+}
+
+func ParseInstanceExternalRef(externalRef string) (parent *InstanceParent, resourceID string, err error) {
+	if !strings.HasPrefix(externalRef, serviceDomain) {
+		return nil, "", fmt.Errorf("externalRef should have prefix %s, got %s", serviceDomain, externalRef)
+	}
+	path := strings.TrimPrefix(externalRef, serviceDomain+"/")
+	return ParseInstanceExternal(path)
+}
+
+func ParseInstanceExternal(external string) (parent *InstanceParent, resourceID string, err error) {
```

Comment:
This two functions should not exist in Identity.

Reply:
Moved to instance_reference.go